src/main/java/net/sf/mzmine/parameters/parametertypes/selectors/ScanSelection.java

```diff
@@ -101,112 +96,78 @@ public String getScanDefinition() {
   public Scan[] getMatchingScans(RawDataFile dataFile) {
 
     final List<Scan> matchingScans = new ArrayList<>();
-    boolean offsetSet = false;
-    int offset = 1;
 
     int scanNumbers[] = dataFile.getScanNumbers();
     for (int scanNumber : scanNumbers) {
 
       Scan scan = dataFile.getScan(scanNumber);
-
-      if ((msLevel != null) && (!msLevel.equals(scan.getMSLevel())))
-        continue;
-
-      if ((polarity != null) && (!polarity.equals(scan.getPolarity())))
-        continue;
-
-      if ((spectrumType != null) && (!spectrumType.equals(scan.getSpectrumType())))
-        continue;
-
-      if ((scanNumberRange != null) && (!scanNumberRange.contains(scanNumber))) {
-        continue;
-      } else {
-        if (!offsetSet) {
-          offsetSet = true;
-          offset = scanNumber;
-        }
-      }
-
-      if ((baseFilteringInteger != null) && ((scanNumber - offset) % baseFilteringInteger != 0))
-        continue;
-
-      if ((scanRTRange != null) && (!scanRTRange.contains(scan.getRetentionTime())))
-        continue;
-
-      if (!Strings.isNullOrEmpty(scanDefinition)) {
-
-        final String actualScanDefition = scan.getScanDefinition();
-
-        if (Strings.isNullOrEmpty(actualScanDefition))
-          continue;
-
-        final String regex = TextUtils.createRegexFromWildcards(scanDefinition);
-
-        if (!actualScanDefition.matches(regex))
-          continue;
-
-      }
-
-      matchingScans.add(scan);
+      if (matches(scan))
+        matchingScans.add(scan);
     }
 
     return matchingScans.toArray(new Scan[matchingScans.size()]);
-
   }
 
   public int[] getMatchingScanNumbers(RawDataFile dataFile) {
 
     final List<Integer> matchingScans = new ArrayList<>();
 
     int scanNumbers[] = dataFile.getScanNumbers();
-    boolean offsetSet = false;
-    int offset = 1;
-    for (int scanNumber : scanNumbers) {
 
+    for (int scanNumber : scanNumbers) {
       Scan scan = dataFile.getScan(scanNumber);
+      if (matches(scan))
+        matchingScans.add(scanNumber);
+    }
 
-      if ((msLevel != null) && (!msLevel.equals(scan.getMSLevel())))
-        continue;
+    return Ints.toArray(matchingScans);
+  }
 
-      if ((polarity != null) && (!polarity.equals(scan.getPolarity())))
-        continue;
 
-      if ((spectrumType != null) && (!spectrumType.equals(scan.getSpectrumType())))
-        continue;
+  public boolean matches(Scan scan) {
+    // scan offset was changed
+    int offset = scanNumberRange == null ? 1 : scanNumberRange.lowerEndpoint();
```

Comment:
Actually in case scanNumberRange==null, I think the offset should not be set to 1, but to the first scan number in the file. That is not necessarily 1.

Reply:
In that case, I have changed it to 
offset = scan.getDataFile().getScanNumbers()[0];